tools/smarterr-parser/main.go

```diff
@@ -0,0 +1,299 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+type FileResult struct {
+	FilePath   string
+	IsResource bool
+	Errors     []ErrorMatch
+}
+
+type ErrorMatch struct {
+	LineNumber int
+	Line       string
+	Pattern    string
+}
+
+func main() {
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <folder_path> [--fix]\n", os.Args[0])
+		os.Exit(1)
+	}
+
+	folderPath := os.Args[1]
+	autoFix := len(os.Args) > 2 && os.Args[2] == "--fix"
+
+	results, err := analyzeFolder(folderPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error analyzing folder: %v\n", err)
+		os.Exit(1)
+	}
+
+	if autoFix {
+		err = fixFiles(results)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error fixing files: %v\n", err)
+			os.Exit(1)
+		}
+		fmt.Printf("Fixed %d files with bare return statements\n", countFilesWithBareReturns(results))
+	} else {
+		printResults(results)
+	}
+}
+
+// analyzeFolder walks through the folder and analyzes all .go files
+func analyzeFolder(folderPath string) ([]FileResult, error) {
+	var results []FileResult
+
+	err := filepath.Walk(folderPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !strings.HasSuffix(path, ".go") {
+			return nil
+		}
+
+		result, err := analyzeFile(path)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Warning: Error analyzing file %s: %v\n", path, err)
+			return nil
+		}
+
+		if result.IsResource && len(result.Errors) > 0 {
+			results = append(results, result)
+		}
+
+		return nil
+	})
+
+	return results, err
+}
+
+func analyzeFile(filePath string) (FileResult, error) {
+	result := FileResult{
+		FilePath: filePath,
+		Errors:   []ErrorMatch{},
+	}
+
+	file, err := os.Open(filePath)
+	if err != nil {
+		return result, err
+	}
+	defer func() {
+		_ = file.Close()
+	}()
+
+	scanner := bufio.NewScanner(file)
+	lineNumber := 0
+
+	for scanner.Scan() {
+		lineNumber++
+		line := scanner.Text()
+
+		if isResourceOrDataSource(line) {
+			result.IsResource = true
+			break
+		}
+	}
+
+	if !result.IsResource {
+		return result, nil
+	}
```

Comment:
Are there cases where we would want to use `smarterr` in a non-resource/data source file? One example might be logic shared between multiple resources in a service. 

Is the assumption that the upstream call sites in each resource will wrap the error instead?